### PR TITLE
UI Configuration Service Configuration

### PR DESCRIPTION
**Pull request by lkiesow**
This patch moves the configuration of the user interface configuration
service (providing the `/ui-config` endpoint) from the global
configuration to a service configuration file.

It also adds support for returning an X-Accel-Redirect header along the
way.

### Your pull request should…

* [x] have a concise title
* [x] [close an accompanying issue](https://help.github.com/en/articles/closing-issues-using-keywords) if one exists
* [x] [be against the correct branch](https://docs.opencast.org/develop/developer/development-process#acceptance-criteria-for-patches-in-different-versions)
* [x] include migration scripts and documentation, if appropriate
* [x] pass automated tests
* [x] have a clean commit history
* [x] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)

### CODE REVIEW

File: etc/org.opencastproject.uiconfig.UIConfigRest.cfg [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Directory user interface configuration is served from.
+# Default: ${karaf.etc}/ui-config
+#org.opencastproject.uiconfig.folder=${karaf.etc}/ui-config
+
+# Makes the UI configuration service return an X-Accel-Redirect header using the
+# following path prefix instead of serving the actual file.
+# More details: https://www.nginx.com/resources/wiki/start/topics/examples/x-accel/
+# Default: null
+#x.accel.redirect=/protected-ui-config
```
Comment (KatrinIhler) [inline etc/org.opencastproject.uiconfig.UIConfigRest.cfg:9]: Wouldn't it be nicer for the value in the comment to align with the default value?

File: modules/user-interface-configuration/src/main/java/org/opencastproject/uiconfig/UIConfigRest.java [hunk 2]
```diff
@@ -69,48 +74,65 @@
...
   /** Configuration key for the ui config folder */
   static final String UI_CONFIG_FOLDER_PROPERTY = "org.opencastproject.uiconfig.folder";
+  static final String X_ACCEL_REDIRECT_PROPERTY = "x.accel.redirect";
 
   /** Default Path for the ui configuration folder (relative to ${karaf.etc}) */
   private static final String UI_CONFIG_FOLDER_DEFAULT = "ui-config";
 
   /** The currently used path to the configuration folder */
   private String uiConfigFolder = "";
+  private String xAccelRedirect = null;
```
Comment (KatrinIhler) [inline modules/user-interface-configuration/src/main/java/org/opencastproject/uiconfig/UIConfigRest.java:99]: Not your fault, but maybe we could set uiConfigFolder to null as well?

File: modules/user-interface-configuration/src/test/java/org/opencastproject/uiconfig/UIConfigTest.java [hunk 3]
```diff
@@ -69,49 +71,39 @@ public void setUp() throws Exception {
 
   @Test
   public void testActivate() throws Exception {
+    Map<String, String> properties = EasyMock.createMock(Map.class);
+    EasyMock.expect(properties.get(UI_CONFIG_FOLDER_PROPERTY)).andReturn(null).times(2);
+    EasyMock.expect(properties.get(UI_CONFIG_FOLDER_PROPERTY)).andReturn("/xy").once();
+    EasyMock.expect(properties.get(X_ACCEL_REDIRECT_PROPERTY)).andReturn(null).times(2);
```
Comment (KatrinIhler) [inline modules/user-interface-configuration/src/test/java/org/opencastproject/uiconfig/UIConfigTest.java]: Maybe add a small test case for the new redirect functionality?

File: modules/user-interface-configuration/src/main/java/org/opencastproject/uiconfig/UIConfigRest.java [hunk 4]
```diff
@@ -69,48 +74,65 @@
...
-
-      if (StringUtils.isBlank(karafetc)) {
-        throw new ConfigurationException(UI_CONFIG_FOLDER_PROPERTY + " not set and unable to"
-                                         + " fall back to default location based on ${karaf.etc}");
+  @Activate
+  @Modified
+  public void activate(BundleContext context, Map<String, String> properties)
+          throws ConfigurationException, IOException {
+    uiConfigFolder = properties.get(UI_CONFIG_FOLDER_PROPERTY);
+    logger.debug("UI configuration folder set to '{}'", uiConfigFolder);
```
Comment (KatrinIhler) [inline modules/user-interface-configuration/src/main/java/org/opencastproject/uiconfig/UIConfigRest.java]: Do we really need this if we always log the canonical path further down? If we do, maybe we should differentiate the logging message a bit.

**Review by KatrinIhler — COMMENTED** [review]
Some nitpicking, but it looks good overall. :)

File: etc/org.opencastproject.uiconfig.UIConfigRest.cfg (hunk 1 above)
Comment (lkiesow) [reply etc/org.opencastproject.uiconfig.UIConfigRest.cfg:9]: No, it wouldn't. This pattern is used all over Opencast's configuration file and I really like it.
This way, you can clearly differentiate between documentation and commented out option.

File: modules/user-interface-configuration/src/main/java/org/opencastproject/uiconfig/UIConfigRest.java (hunk 4 above)
Comment (lkiesow) [reply modules/user-interface-configuration/src/main/java/org/opencastproject/uiconfig/UIConfigRest.java]: Here you see what the actual configured value is while magic is involved in the other log further down. They are both debug level logs, and you wouldn't normally see them. But I don't care that much, and I can remove either one if you want me to.

File: modules/user-interface-configuration/src/test/java/org/opencastproject/uiconfig/UIConfigTest.java (hunk 3 above)
Comment (lkiesow) [reply modules/user-interface-configuration/src/test/java/org/opencastproject/uiconfig/UIConfigTest.java]: done

File: modules/user-interface-configuration/src/main/java/org/opencastproject/uiconfig/UIConfigRest.java (hunk 2 above)
Comment (lkiesow) [reply modules/user-interface-configuration/src/main/java/org/opencastproject/uiconfig/UIConfigRest.java:99]: done

File: etc/org.opencastproject.uiconfig.UIConfigRest.cfg (hunk 1 above)
Comment (KatrinIhler) [reply etc/org.opencastproject.uiconfig.UIConfigRest.cfg:9]: Hm, good point. Let's leave it as is.

File: modules/user-interface-configuration/src/main/java/org/opencastproject/uiconfig/UIConfigRest.java (hunk 4 above)
Comment (KatrinIhler) [reply modules/user-interface-configuration/src/main/java/org/opencastproject/uiconfig/UIConfigRest.java]:
```suggestion
    logger.debug("UI configuration folder configured as '{}'", uiConfigFolder);
```
?

**Review by KatrinIhler — APPROVED** [review]

**Pull request merged**